tests/e2e/lib.sh

```diff
@@ -576,28 +576,144 @@ check_for_errors_in_stackrox_logs() {
-        local check_out=""
+    local pod_objects=()
+    _get_pod_objects "${dir}"
+    _verify_item_count "${dir}" "${#pod_objects[*]}"
+
+    declare -A podnames_by_app
+    _group_pods_by_app_label
+
+    # Check the logs for each app separately
+    local app logs check_out summary
+    local failure_found="false"
+    LOGCHECK_SCRIPT="${LOGCHECK_SCRIPT:-scripts/ci/logcheck/check.sh}"
+    for app in "${!podnames_by_app[@]}"; do
+        logs="$(_get_logs_for_app "${app}")"
         # shellcheck disable=SC2086
-        if ! check_out="$(scripts/ci/logcheck/check.sh $filtered)"; then
-            local last_log="${check_out##*.log:}"
-            save_junit_failure "SuspiciousLog" "${last_log##*: }" "$check_out"
-            die "ERROR: Found at least one suspicious log file entry."
+        if [[ -n "${logs}" ]] && ! check_out="$(${LOGCHECK_SCRIPT} ${logs})"; then
+            summary="$(summarize_check_output "${check_out}")"
+            save_junit_failure "SuspiciousLog-${app}" "${summary}" "$check_out"
+            failure_found="true"
         else
-            save_junit_success "SuspiciousLog" "Suspicious entries in log file(s)"
+            save_junit_success "SuspiciousLog-${app}" "Suspicious entries in log file(s)"
+        fi
+    done
+    if [[ "${failure_found}" == "true" ]]; then
+        die "ERROR: Found at least one suspicious log file entry."
+    fi
+}
+
+_get_pod_objects() {
+    local dir="$1"
+
+    local nullglob_setting
+    nullglob_setting="$(shopt nullglob || true)"
+    if [[ "${nullglob_setting}" =~ off ]]; then
+        shopt -s nullglob
+    fi
+
+    local pod
+    for pod in "${dir}"/*_object.json; do
+        pod_objects[${#pod_objects[*]}]="${pod}"
+    done
+
+    if [[ "${nullglob_setting}" =~ off ]]; then
+        shopt -u nullglob
+    fi
+}
+
+_verify_item_count() {
+    local dir="$1"
+    local pod_object_count="$2"
+
+    # ITEM_COUNT.txt is used to keep this function in sync with the file output
+    # used by ./scripts/ci/collect-service-logs.sh
+
+    if [[ ! -f "${dir}/ITEM_COUNT.txt" ]]; then
+        die "ITEM_COUNT.txt is missing. (Check output from ./scripts/ci/collect-service-logs.sh"
+    fi
+
+    local item_count
+    item_count="$(cat "${dir}/ITEM_COUNT.txt")"
+
+    if [[ "${item_count}" != "${pod_object_count}" ]]; then
+        die "The recorded number of items (${item_count}) differs from the objects found (${pod_object_count})"
+    fi
+}
+
+_group_pods_by_app_label() {
+    local pod_object app podname
+    for pod_object in "${pod_objects[@]}"; do
+        podname="$(jq -r '.metadata.name' < "${pod_object}")"
+        if [[ -z "${podname}" || "${podname}" == "null" ]]; then
+            die "ERROR: All pods should have a name! (check ${podname})"
+        fi
+        app="$(jq -r '.metadata.labels.app' < "${pod_object}")"
+        if [[ -z "${app}" || "${app}" == "null" ]]; then
+            die "ERROR: All Stackrox pods should have an app label (check ${podname})"
```

Comment:
I think it would be more robust here to use an "unknown" app rather than crash early.
If you feel strongly about this, then return an error when exiting if such unknowns were found.

Reply:
If a pod appears in the `stackrox` namespace without a `metadata.labels.app` then it is a bug introduced by the changes on a PR so I'd like to surface it as a failure to the PR author. Otherwise this is a leaky assumption that may no longer hold true over time.

Reply:
Sure. But it should not prevent a debugging script from completing first, if at all possible, right?